xonsh/history.py

```diff
@@ -407,103 +406,84 @@ def _hist_create_parser():
     return p
 
 
-def _hist_show(ns=None, hist=None, start_index=None, end_index=None,
-               start_time=None, end_time=None, location=None):
-    """Show the requested portion of shell history.
-    Accepts multiple history sources (xonsh, bash, zsh)
-
-    May be invoked as an alias with history all/bash/zsh which will
-    provide history as stdout or with __xonsh_history__.show()
-    which will return the history as a list with each item
-    in the tuple form (name, start_time, index).
+def _hist_get_session(session='session', location=None):
+    "Yield commands from history."
+    cmds = _HIST_SESSIONS[session](location=location)
+    try:
+        yield from cmds
+    except TypeError:
+        pass
+
+
+def _hist_get_portion(commands, slices):
+    "Yield from portions of history commands. "
+    commands = list(commands)
```

Comment:
OK, sure, but we *do* know the length of the commands at one point because everything in `_HIST_SESSIONS` returns a list.  You can then use `slice.indicies()` to get positive indicies based on the length of the commands.  

I guess my point is that `_hist_get_session()` gets a list and then turns it into a iterator.  This function then gets an iterator and turns it back into a list with the same contents.  That seems a little excessive.  It would be better to just use the original list directly or have `_hist_get_session()` return the list it sees.